servant-server/src/Servant/Server/Internal.hs

```diff
@@ -718,8 +720,8 @@ instance ( AllCTUnrender list a, HasServer sublayout
       mrqbody <- handleCTypeH (Proxy :: Proxy list) (cs contentTypeH)
              <$> lazyRequestBody request
       case mrqbody of
-        Nothing -> return $ failWith $ UnsupportedMediaType
-        Just (Left e) -> return $ failWith $ InvalidBody e
+        Nothing -> return $ FailFatal err415
+        Just (Left e) -> return $ FailFatal err400 { errBody = cs e }
```

Comment:
Ok, getting rid of request body memoization would be nice indeed. But the price seems high.